### ISSUE TITLE
[💡SUG] Implement more negative sampling loss functions

### ISSUE THREAD

**Issue by deklanw** [RUCAIBox/RecBole#583]
LightFM, in addition to BPR, implements WARP and k-OS WARP loss. They find that it outperforms BPR https://making.lyst.com/lightfm/docs/examples/warp_loss.html (performance comparison, citations to relevant papers, etc in this article)

The LightGCN papers says,

> We are aware of other advanced negative sampling strategies which might improve the LightGCN training, such as the hard negative sampling [31] and adversarial sampling [9]. We leave this extension in the future since it is not the focus of this work.

Citation [31] is "Improving Pairwise Learning for Item Recommendation from Implicit Feedback" by Rendle wherein he suggests the Adaptive Oversampling loss and demonstrates its improvement over BPR. (This paper is also cited in the LightFM docs above as similar to WARP). At the end of the paper, he suggests his method might be faster than WARP (but doesn't compare):

![image](https://user-images.githubusercontent.com/11601778/102233460-4955cc00-3ebe-11eb-85a6-381347047afa.png)

These losses could of course be applied to many models, but my suggestion is that applying them to matrix factorization would be a good start. The "BPR" model could be renamed to "MF" and different loss functions could be passed as hyperparameters.

**Comment by batmanfly** [comment]
> LightFM, in addition to BPR, implements WARP and k-OS WARP loss. They find that it outperforms BPR https://making.lyst.com/lightfm/docs/examples/warp_loss.html (performance comparison, citations to relevant papers, etc in this article)
> 
> The LightGCN papers says,
> 
> > We are aware of other advanced negative sampling strategies which might improve the LightGCN training, such as the hard negative sampling [31] and adversarial sampling [9]. We leave this extension in the future since it is not the focus of this work.
> 
> Citation [31] is "Improving Pairwise Learning for Item Recommendation from Implicit Feedback" by Rendle wherein he suggests the Adaptive Oversampling loss and demonstrates its improvement over BPR. (This paper is also cited in the LightFM docs above as similar to WARP). At the end of the paper, he suggests his method might be faster than WARP (but doesn't compare):
> 
> ![image](https://user-images.githubusercontent.com/11601778/102233460-4955cc00-3ebe-11eb-85a6-381347047afa.png)
> 
> These losses could of course be applied to many models, but my suggestion is that applying them to matrix factorization would be a good start. The "BPR" model could be renamed to "MF" and different loss functions could be passed as hyperparameters.

@deklanw Very nice suggestion!  As you mentioned, there are several sampling strategies proposed recently, which may be effective than BPR loss. We are considering adding more optimization tricks or methods in the coming version.  Some  sampling algorithms might require significant revision of underlying data structure in our library, and will take a long time. We will try our best to  implement them. Will update you at this issue page if there were new changes in our library. 

For renaming the "BPR" model, a major consideration is that the name "BPR" has been widely used in the literature for referring to the Rendle's method. Therefore, we will consider renameing "BPR" as "BPRMF". 

---Wayne Xin Zhao
